dll/cpl/input/settings_page.c

```diff
@@ -10,195 +10,358 @@
-                    }
+    ExtractIconExW(szPath, 0, NULL, &hIconSm, 1);
+    return hIconSm;
+}
 
-                    DeleteObject(hFont);
-                }
-            }
+HTREEITEM FindLanguageInList(HWND hwndList, LPCTSTR pszLangName)
+{
+    TV_ITEM item;
+    TCHAR szText[128];
+    HTREEITEM hItem;
 
-            DeleteObject(hBmpNew);
-        }
+    hItem = TreeView_GetRoot(hwndList);
+    while (hItem)
+    {
+        szText[0] = 0;
+        item.mask       = TVIF_TEXT | TVIF_HANDLE;
+        item.pszText    = szText;
+        item.cchTextMax = _countof(szText);
+        item.hItem      = hItem;
+        TreeView_GetItem(hwndList, &item);
+        if (lstrcmpi(szText, pszLangName) == 0)
+            return hItem;
+
+        hItem = TreeView_GetNextSibling(hwndList, hItem);
     }
 
-    DeleteDC(hdc);
-    DeleteObject(hBitmap);
-
-    return hIcon;
+    return NULL;
 }
 
-
 static VOID
-SetControlsState(HWND hwndDlg)
+AddToInputListView(HWND hwndList, INPUT_LIST_NODE *pInputNode)
 {
-    HWND hwndList = GetDlgItem(hwndDlg, IDC_KEYLAYOUT_LIST);
-    INT iSelected = ListView_GetNextItem(hwndList, -1, LVNI_SELECTED);
-    INT nCount = ListView_GetItemCount(hwndList);
-    BOOL bCanRemove = (iSelected != -1) && (nCount >= 2);
-    BOOL bCanDefault = (iSelected != -1) && (nCount >= 2);
-    BOOL bCanProp = (iSelected != -1);
-
-    LV_ITEM item = { LVIF_PARAM, iSelected };
-    if (ListView_GetItem(hwndList, &item))
+    TV_ITEM item;
+    TV_INSERTSTRUCT insert;
+    HIMAGELIST hImageList = TreeView_GetImageList(hwndList, TVSIL_NORMAL);
+    WCHAR szKeyboard[64];
+    HTREEITEM hItem;
+    BOOL bBold = !!(pInputNode->wFlags & INPUT_LIST_NODE_FLAG_DEFAULT);
+
+    hItem = FindLanguageInList(hwndList, pInputNode->pLocale->pszName);
+    if (hItem == NULL)
     {
-        INPUT_LIST_NODE *pInput = (INPUT_LIST_NODE*)item.lParam;
+        // Language icon
+        INT LangImageIndex = -1;
+        HICON hLangIcon = CreateLayoutIcon(LOWORD(pInputNode->pLocale->dwId));
+        if (hLangIcon)
+        {
+            LangImageIndex = ImageList_AddIcon(hImageList, hLangIcon);
+            DestroyIcon(hLangIcon);
+        }
 
-        if (pInput && (pInput->wFlags & INPUT_LIST_NODE_FLAG_DEFAULT))
+        // Language
+        ZeroMemory(&item, sizeof(item));
+        item.mask           = TVIF_TEXT | TVIF_IMAGE | TVIF_PARAM | TVIF_SELECTEDIMAGE | TVIF_STATE;
+        item.pszText        = pInputNode->pLocale->pszName;
+        item.iImage         = LangImageIndex;
+        item.iSelectedImage = LangImageIndex;
+        item.lParam         = LOWORD(pInputNode->pLocale->dwId); // HIWORD(item.lParam) == 0
+        if (bBold)
         {
-            bCanDefault = FALSE;
+            item.state = item.stateMask = TVIS_BOLD;
         }
+        insert.hParent      = TVI_ROOT;
+        insert.hInsertAfter = TVI_LAST;
+        insert.item         = item;
+        hItem = TreeView_InsertItem(hwndList, &insert);
+
+        // Keyboard
```

Comment:
Note: This is not necessarily always "Keyboard": indeed you can have "Other" (or maybe some other description): this is the "type" of the input method. "Other" is for example used for pen-driven inputs.

Reply:
Comments fixed.